{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "82882c9b-d218-46b1-9883-36963ff7aae4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50db5aa2",
   "metadata": {},
   "source": [
    "Welcome to `deferential_expression`. The purpose of this package is to give Python users easy broad access to differential expression libraries in R.\n",
    "\n",
    "Using R-based libraries from Python can have a significant conversion overhead, depending on the size of this matrix. We solve this problem by converting expression to R once using `rpy2`, and providing access to R-based differential expression packages for these matrices.\n",
    "\n",
    "Internally, we use our class `RMatrixAdapter`, to provide `numpy`-like behaviour for an R-matrix. Users, however, never needs to seeds to see `RMatrixAdapter`, storing their expression data in a BiocPy `SummarizedExperiment` and using a simple one-off conversion."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3158c6d",
   "metadata": {},
   "source": [
    "## Installation\n",
    "For installation there are two main ways: (1) conda-based and (2) using System-installed R.\n",
    "\n",
    "In the first case, you may use the provided `environment.yaml` file. This will install all required dependencies, including Python, R, `rpy2` and the R-packages, within a conda environment.\n",
    "\n",
    "If you don´t want to use a conda-installed R, a system-installed R installation may be used. R needs to be added to `PATH`, and `rpy2` installed. `deferential_expression` can then be installed using \n",
    "`\"deferential_expression @ git+https://github.com/MaximilianNuber/deferential_expression.git@main\"`.\n",
    "As for the R dependencies, using R functionalities for the first time will trigger a check, and missing dependencies will be installed automatically.\n",
    "\n",
    "Alternatively, the repository can be cloned and installed with\n",
    "```\n",
    "pip install -e .\n",
    "# (optional) \n",
    "# pip install -e \".[docs]\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb832fc3",
   "metadata": {},
   "source": [
    "## Quickstart: recount3 ➜ edgeR (all from Python)\n",
    "\n",
    "But let´s get to an example of how this package can be used.\n",
    "\n",
    "The snippet below runs R code from Python to download SRP149978 with recount3, extracts raw counts + sample metadata, builds an R-backed container, and runs an edgeR quasi-likelihood (QL) pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1113d7ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "# Setting up the data and converting from R.\n",
    "\n",
    "import os\n",
    "os.environ[\"R_HOME\"] = \"/Users/maximiliannuber/miniconda3/envs/def_exp_test_env/lib/R\"\n",
    "\n",
    "import pandas as pd\n",
    "from bioc2ri.lazy_r_env import r  # lazy rpy2 env, see \"bioc2ri\"\n",
    "import deferential_expression as de\n",
    "\n",
    "# --- 1) Use recount3 in R to pull a moderate-sized project and make a 'condition' factor\n",
    "r.ro.r(\"\"\"\n",
    "\n",
    "suppressMessages(library(recount3))\n",
    "suppressMessages(library(stringr))\n",
    "\n",
    "human_projects <- available_projects()\n",
    "my_project <- subset(human_projects, project == \"SRP149978\" & project_type == \"data_sources\")\n",
    "rse_gene <- create_rse(my_project)\n",
    "\n",
    "rse_gene$condition <- rse_gene$sra.experiment_title |>\n",
    "  stringr::str_split(\":\") |> sapply(function(x) x[2]) |>\n",
    "  stringr::str_split(\";\") |> sapply(function(x) x[1]) |>\n",
    "  stringr::str_split(\"_\") |> sapply(function(x) x[1]) |>\n",
    "  stringr::str_replace_all(\" \", \"\") |>\n",
    "  as.factor() |> relevel(\"Control\")\n",
    "\"\"\")\n",
    "\n",
    "# --- 2) Pull counts (R matrix, stays R-backed) and colData (to pandas)\n",
    "counts   = r.r2py(r.ro.r(\"assay(rse_gene, 'raw_counts')\"))\n",
    "coldata_df = r.r2py(r.ro.r(\"as.data.frame(colData(rse_gene))\"))\n",
    "rowdata_df = r.r2py(r.ro.r(\"as.data.frame(rowData(rse_gene))\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d8ccc81",
   "metadata": {},
   "source": [
    "We use the `recount3` Bioconductor package to get data from R. We have a `numpy` RNA-seq count matrix, sample-wise meta data, and feature wise meta data. \n",
    "\n",
    "Gene expression matrices in Bioconductor are stored in objects like `SummarizedExperiment` in the format `(genes x samples)`.\n",
    "\n",
    "First, we store our data in `SummarizedExperiment` Python-only:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b1e85668",
   "metadata": {},
   "outputs": [],
   "source": [
    "from summarizedexperiment import SummarizedExperiment\n",
    "\n",
    "se = SummarizedExperiment(\n",
    "    assays = {\"counts\": counts},\n",
    "    column_data = coldata_df,\n",
    "    row_data = rowdata_df\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00c5b920",
   "metadata": {},
   "source": [
    "A common bottleneck, especially for large expression matrices, is the conversion from Python to R. We want to perform this conversion only once. In the following snippet, we construct an `RESummarizedExperiment`, which converts matrices in `assays` to R-matrices. Otherwise, it behaves like a BiocPy `SummarizedExperiment`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "766430a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rpy2.robjects.vectors.FloatMatrix"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rse = de.RESummarizedExperiment.from_summarized_experiment(se)\n",
    "\n",
    "# take a look at the assay:\n",
    "type(rse.assay_r(\"counts\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c24b26c-4103-47be-becd-3c134906285d",
   "metadata": {},
   "source": [
    "As you can see, the numpy array from SummarizedExperiment has been converted to an R-matrix.\n",
    "\n",
    "Dedicated functions can use this R-matrix and enrich the `RESummarizedExperiment` with further assays from R libraries. Let´s test the `edgeR`-workflow:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "15441499-c679-4715-b1bf-d753ef34a288",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True     32222\n",
      "False    31634\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count    92.000000\n",
       "mean      1.004803\n",
       "std       0.097194\n",
       "min       0.730631\n",
       "25%       0.934352\n",
       "50%       1.011183\n",
       "75%       1.075588\n",
       "max       1.211238\n",
       "Name: norm.factors, dtype: float64"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# edgeR-preprocessing:\n",
    "\n",
    "keep = de.edger.filter_by_expr(rse, group = rse.col_data[\"condition\"])\n",
    "print(\n",
    "    pd.Series(keep).value_counts()\n",
    ")\n",
    "\n",
    "rse = rse[keep, :]\n",
    "\n",
    "rse = de.edger.calc_norm_factors(rse)\n",
    "\n",
    "rse.col_data.to_pandas()[\"norm.factors\"].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c258b41-3eaf-4db3-b20a-70d39d0e2453",
   "metadata": {},
   "source": [
    "A small explanation of what happened: `de.edger.filter_by_expr` calls the R function `edgeR::filterByExpr` on the specified assay, by default \"counts\", and the output is automatically converted into a numpy-array and returned. Using this boolean array, we can simply subset an ´RESummarizedExperiment`. \n",
    "\n",
    "`de.edger.calc_norm_factors` calls `edger::calcNormFactors`, and adds the \"norm.factors\" column to the column data. All arguments can be specified and are forwarded to the R-function.\n",
    "\n",
    "\n",
    "Next up, we want to use the edgeR model. First, we need to construct a design matrix. \n",
    "We do this in Python with current tools, we recommend `formulaic_contrasts`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e7b45722-8d0a-41fc-bee2-178cdf862484",
   "metadata": {},
   "outputs": [],
   "source": [
    "from formulaic_contrasts import FormulaicContrasts\n",
    "\n",
    "fc = FormulaicContrasts(rse.col_data.to_pandas(), \"~condition\")\n",
    "design = fc.design_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "29f9c2b1-7b11-411c-b072-eb02e0dfaab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = de.edger.glm_ql_fit(rse, design = design)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bf2289a-1f68-4de8-ae9c-3bc9b14891ae",
   "metadata": {},
   "source": [
    "The design matrix is a pandas dataframe and can be treated as such. We pass the `RESummarizedExperiment` and design to the fit-function, and get a Python-object containing the edgeR fit-object.\n",
    "\n",
    "This fit-object can be passed to the test function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a5b21931-e775-4bec-8cb7-42dcf7a0d0ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = de.edger.glm_ql_ftest(fit, coef = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ceaa9d44-6a47-4bf4-a19f-425ad6124846",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>logCPM</th>\n",
       "      <th>F</th>\n",
       "      <th>PValue</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.5</th>\n",
       "      <td>0.149352</td>\n",
       "      <td>30.142036</td>\n",
       "      <td>0.538628</td>\n",
       "      <td>0.464784</td>\n",
       "      <td>0.661409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267.1</th>\n",
       "      <td>0.664313</td>\n",
       "      <td>26.728468</td>\n",
       "      <td>9.757847</td>\n",
       "      <td>0.002373</td>\n",
       "      <td>0.032758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.5</th>\n",
       "      <td>0.265684</td>\n",
       "      <td>33.449910</td>\n",
       "      <td>9.872641</td>\n",
       "      <td>0.002229</td>\n",
       "      <td>0.031747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.5</th>\n",
       "      <td>-0.038612</td>\n",
       "      <td>24.673589</td>\n",
       "      <td>0.013832</td>\n",
       "      <td>0.906534</td>\n",
       "      <td>0.952470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945.1</th>\n",
       "      <td>0.014600</td>\n",
       "      <td>28.045229</td>\n",
       "      <td>0.005973</td>\n",
       "      <td>0.938558</td>\n",
       "      <td>0.969457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000279274.2</th>\n",
       "      <td>-0.218581</td>\n",
       "      <td>26.507425</td>\n",
       "      <td>2.391306</td>\n",
       "      <td>0.125375</td>\n",
       "      <td>0.307094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228786.5</th>\n",
       "      <td>-0.080047</td>\n",
       "      <td>25.265041</td>\n",
       "      <td>0.069426</td>\n",
       "      <td>0.792641</td>\n",
       "      <td>0.887383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000229238.3</th>\n",
       "      <td>-0.555580</td>\n",
       "      <td>24.248539</td>\n",
       "      <td>2.084508</td>\n",
       "      <td>0.150467</td>\n",
       "      <td>0.341842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000231514.1</th>\n",
       "      <td>-0.225417</td>\n",
       "      <td>26.020608</td>\n",
       "      <td>0.937149</td>\n",
       "      <td>0.335408</td>\n",
       "      <td>0.546330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237917.1</th>\n",
       "      <td>-0.075558</td>\n",
       "      <td>26.942928</td>\n",
       "      <td>0.053453</td>\n",
       "      <td>0.817665</td>\n",
       "      <td>0.901577</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32222 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      logFC     logCPM         F    PValue       FDR\n",
       "ENSG00000223972.5  0.149352  30.142036  0.538628  0.464784  0.661409\n",
       "ENSG00000278267.1  0.664313  26.728468  9.757847  0.002373  0.032758\n",
       "ENSG00000227232.5  0.265684  33.449910  9.872641  0.002229  0.031747\n",
       "ENSG00000243485.5 -0.038612  24.673589  0.013832  0.906534  0.952470\n",
       "ENSG00000239945.1  0.014600  28.045229  0.005973  0.938558  0.969457\n",
       "...                     ...        ...       ...       ...       ...\n",
       "ENSG00000279274.2 -0.218581  26.507425  2.391306  0.125375  0.307094\n",
       "ENSG00000228786.5 -0.080047  25.265041  0.069426  0.792641  0.887383\n",
       "ENSG00000229238.3 -0.555580  24.248539  2.084508  0.150467  0.341842\n",
       "ENSG00000231514.1 -0.225417  26.020608  0.937149  0.335408  0.546330\n",
       "ENSG00000237917.1 -0.075558  26.942928  0.053453  0.817665  0.901577\n",
       "\n",
       "[32222 rows x 5 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "82d98fa5-7684-4904-a406-b1c6e9ca3563",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>bp_length</th>\n",
       "      <th>phase</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>tag</th>\n",
       "      <th>rownames</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.5</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1735.0</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267.1</th>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>68.0</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>3</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.5</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1351.0</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.5</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1021.0</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>2</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945.1</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1319.0</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>ENSG00000239945.1</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>RP11-34P13.8</td>\n",
       "      <td>2</td>\n",
       "      <td>OTTHUMG00000001097.2</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>ENSG00000239945.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000279274.2</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>75.0</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>ENSG00000279274.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>RP11-533E23.2</td>\n",
       "      <td>2</td>\n",
       "      <td>OTTHUMG00000192107.1</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>ENSG00000279274.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228786.5</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1192.0</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>ENSG00000228786.5</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>LINC00266-4P</td>\n",
       "      <td>2</td>\n",
       "      <td>OTTHUMG00000045201.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>ENSG00000228786.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000229238.3</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>1671.0</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>ENSG00000229238.3</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>PPP1R12BP1</td>\n",
       "      <td>2</td>\n",
       "      <td>OTTHUMG00000036764.2</td>\n",
       "      <td>NA_character_</td>\n",
       "      <td>ENSG00000229238.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000231514.1</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>640.0</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>ENSG00000231514.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>FAM58CP</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMG00000036813.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>ENSG00000231514.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237917.1</th>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>2337.0</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>ENSG00000237917.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>PARP4P1</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMG00000036812.1</td>\n",
       "      <td>overlapping_locus</td>\n",
       "      <td>ENSG00000237917.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32222 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    source  type  bp_length       phase            gene_id  \\\n",
       "ENSG00000223972.5   HAVANA  gene     1735.0 -2147483648  ENSG00000223972.5   \n",
       "ENSG00000278267.1  ENSEMBL  gene       68.0 -2147483648  ENSG00000278267.1   \n",
       "ENSG00000227232.5   HAVANA  gene     1351.0 -2147483648  ENSG00000227232.5   \n",
       "ENSG00000243485.5   HAVANA  gene     1021.0 -2147483648  ENSG00000243485.5   \n",
       "ENSG00000239945.1   HAVANA  gene     1319.0 -2147483648  ENSG00000239945.1   \n",
       "...                    ...   ...        ...         ...                ...   \n",
       "ENSG00000279274.2   HAVANA  gene       75.0 -2147483648  ENSG00000279274.2   \n",
       "ENSG00000228786.5   HAVANA  gene     1192.0 -2147483648  ENSG00000228786.5   \n",
       "ENSG00000229238.3   HAVANA  gene     1671.0 -2147483648  ENSG00000229238.3   \n",
       "ENSG00000231514.1   HAVANA  gene      640.0 -2147483648  ENSG00000231514.1   \n",
       "ENSG00000237917.1   HAVANA  gene     2337.0 -2147483648  ENSG00000237917.1   \n",
       "\n",
       "                                            gene_type      gene_name level  \\\n",
       "ENSG00000223972.5  transcribed_unprocessed_pseudogene        DDX11L1     2   \n",
       "ENSG00000278267.1                               miRNA      MIR6859-1     3   \n",
       "ENSG00000227232.5              unprocessed_pseudogene         WASH7P     2   \n",
       "ENSG00000243485.5                             lincRNA    MIR1302-2HG     2   \n",
       "ENSG00000239945.1                             lincRNA   RP11-34P13.8     2   \n",
       "...                                               ...            ...   ...   \n",
       "ENSG00000279274.2                processed_pseudogene  RP11-533E23.2     2   \n",
       "ENSG00000228786.5                             lincRNA   LINC00266-4P     2   \n",
       "ENSG00000229238.3              unprocessed_pseudogene     PPP1R12BP1     2   \n",
       "ENSG00000231514.1                processed_pseudogene        FAM58CP     1   \n",
       "ENSG00000237917.1              unprocessed_pseudogene        PARP4P1     1   \n",
       "\n",
       "                            havana_gene                tag           rownames  \n",
       "ENSG00000223972.5  OTTHUMG00000000961.2      NA_character_  ENSG00000223972.5  \n",
       "ENSG00000278267.1         NA_character_      NA_character_  ENSG00000278267.1  \n",
       "ENSG00000227232.5  OTTHUMG00000000958.1      NA_character_  ENSG00000227232.5  \n",
       "ENSG00000243485.5  OTTHUMG00000000959.2         ncRNA_host  ENSG00000243485.5  \n",
       "ENSG00000239945.1  OTTHUMG00000001097.2  overlapping_locus  ENSG00000239945.1  \n",
       "...                                 ...                ...                ...  \n",
       "ENSG00000279274.2  OTTHUMG00000192107.1      NA_character_  ENSG00000279274.2  \n",
       "ENSG00000228786.5  OTTHUMG00000045201.1  overlapping_locus  ENSG00000228786.5  \n",
       "ENSG00000229238.3  OTTHUMG00000036764.2      NA_character_  ENSG00000229238.3  \n",
       "ENSG00000231514.1  OTTHUMG00000036813.1  overlapping_locus  ENSG00000231514.1  \n",
       "ENSG00000237917.1  OTTHUMG00000036812.1  overlapping_locus  ENSG00000237917.1  \n",
       "\n",
       "[32222 rows x 11 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rse.row_data.to_pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "df6c7843-67f9-46ca-b20b-e7cefaa81aa5",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'set_row_names' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)\n",
      "Cell \u001b[0;32mIn[12], line 1\u001b[0m\n",
      "\u001b[0;32m----> 1\u001b[0m rse \u001b[38;5;241m=\u001b[39m \u001b[43mrse\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mset_row_names\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrse\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrow_data\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene_name\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n",
      "\n",
      "File \u001b[0;32m~/projects/my_packages/pyrtools_conversions/deferential_expression/src/deferential_expression/resummarizedexperiment.py:562\u001b[0m, in \u001b[0;36mRESummarizedExperiment.set_row_names\u001b[0;34m(self, names)\u001b[0m\n",
      "\u001b[1;32m    560\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m assay \u001b[38;5;129;01min\u001b[39;00m assay_names:\n",
      "\u001b[1;32m    561\u001b[0m     mat \u001b[38;5;241m=\u001b[39m assays[assay]\u001b[38;5;241m.\u001b[39mrmat\n",
      "\u001b[0;32m--> 562\u001b[0m     mat_newnames \u001b[38;5;241m=\u001b[39m \u001b[43mset_row_names\u001b[49m(mat, \u001b[38;5;28mlist\u001b[39m(names))\n",
      "\u001b[1;32m    564\u001b[0m     new_assays[assay] \u001b[38;5;241m=\u001b[39m RMatrixAdapter(mat_newnames)\n",
      "\u001b[1;32m    566\u001b[0m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mset_row_names(names)\n",
      "\n",
      "\u001b[0;31mNameError\u001b[0m: name 'set_row_names' is not defined"
     ]
    }
   ],
   "source": [
    "rse = rse.set_row_names(rse.row_data[\"gene_name\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e2d4579c-69ff-4cd2-8596-68e8efe0c431",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bioc2ri.rnames import get_rownames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "71565071-e16c-4c29-9610-8141543f421e",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'set_row_names' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[20]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mrse\u001b[49m\u001b[43m.\u001b[49m\u001b[43mset_row_names\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrse\u001b[49m\u001b[43m.\u001b[49m\u001b[43mrow_data\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mgene_name\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m<string>:9\u001b[39m, in \u001b[36mset_row_names\u001b[39m\u001b[34m(self, names)\u001b[39m\n",
      "\u001b[31mNameError\u001b[39m: name 'set_row_names' is not defined"
     ]
    }
   ],
   "source": [
    "rse.set_row_names(rse.row_data[\"gene_name\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38278b3b-dae3-4e82-a9c4-fff223c2121d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
